### PR TITLE
Updates for WND 2.1.0

### PR DESCRIPTION
**Pull request by mmattel**
These PR are updates for the upcoming WND 2.1.0 release, see https://github.com/owncloud/windows_network_drive/issues/357 (Windows Network Drive 2.1.0)

Fixes: https://github.com/owncloud/docs/issues/3883
Fixes: https://github.com/owncloud/docs/issues/4162
Fixes: https://github.com/owncloud/core/pull/39101 (config_apps_sample changes for WND 2.1.0)

- Reset password on invalid login attempt 343 ✔️ 
- Ldapmembership cache 355 ✔️ (already implemented)
- Adjust WND permissions 351
_(checked, no change needed)_
- Ensure wnd2 cache is cleaned up when actions are performed in the inner 349
_(checked, no change needed)_
- Add support to use the service account's password for WND2 mounts 362 ✔️ 
- Additional check for detecting directories with libsmbclient 4.11+ 366 ✔️

Backport to 10.8 only

@pmaier1 @jvillafanez fyi

### CODE REVIEW

**Review by phil-davis — APPROVED** [review]

**Comment by mmattel** [comment]
Put the PR to draft to protect against accidentally merging.
We will get a notice when WND 2.1.0 is released.

File: modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc [hunk 1]
```diff
@@ -485,6 +485,8 @@ Possible keys: `wnd.activity.sendToSharees` BOOL
 
 Possible keys: `wnd.groupmembership.checkUserFirst` BOOL
 
+*Note* With WND 2.1.0, key `wnd.storage.testForHiddenMount` is obsolete and has been removed completely
```
Comment (EParzefall) [inline modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc:488]:
```suggestion
*Note* With WND 2.1.0, key `wnd.storage.testForHiddenMount` is obsolete and has been removed completely.
```

File: modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc [hunk 2]
```diff
@@ -485,6 +485,8 @@ Possible keys: `wnd.activity.sendToSharees` BOOL
...
+
 === Mandatory listener reconnect to the database
```
Comment (EParzefall) [inline modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc:490]:
```suggestion
=== Mandatory Listener Reconnect to the Database
```

File: modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc [hunk 3]
```diff
@@ -506,19 +508,29 @@ being idle for a long time.
 'wnd.logging.enable' => false,
 ....
 
-=== Check for visible target mount folders when connecting
-Ensure that the connectivity check verifies the mount point is visible.
+=== The way file attributes for folders and files will be handled
```
Comment (EParzefall) [inline modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc:511]:
```suggestion
=== The Way File Attributes for Folders and Files Will Be Handled
```

File: modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc [hunk 4]
```diff
@@ -506,19 +508,29 @@ being idle for a long time.
...
+There are 3 possible values: "none", "stat" and "getxattr":
+
+- "stat". This is the default if the option is missing or has an invalid value
```
Comment (EParzefall) [inline modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc:514]:
```suggestion
- "stat". This is the default if the option is missing or has an invalid value.
```

File: modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc [hunk 5]
```diff
@@ -506,19 +508,29 @@ being idle for a long time.
...
+  This means that the file attributes will be evaluated only for files, NOT for folders
```
Comment (EParzefall) [inline modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc:515]:
```suggestion
  This means that the file attributes will be evaluated only for files, NOT for folders.
```

File: modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc [hunk 6]
```diff
@@ -506,19 +508,29 @@ being idle for a long time.
...
+  Folders will be shown even if the "hidden" file attribute * is set.
+
+- "none". This means that the file attributes won't be evaluated in any case. Both
+  hidden files and folders will be shown, and you can write on read-only files
+  (the action is available in ownCloud, but it will fail in the SMB server).
+
+- "getxattr". This means that file attributes will be evaluated always. However, due to
```
Comment (EParzefall) [inline modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc:522]:
```suggestion
- "getxattr". This means that file attributes will always be evaluated. However, due to
```

File: modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc [hunk 7]
```diff
@@ -506,19 +508,29 @@ being idle for a long time.
...
 
 ==== Code Sample
 
 [source,php]
 ....
-'wnd.storage.testForHiddenMount' => true,
+'wnd.fileInfo.parseAttrs.mode' => stat,
 ....
 
 === Enable or disable the WND in-memory notifier for password changes
```
Comment (EParzefall) [inline modules/admin_manual/pages/configuration/server/config_apps_sample_php_parameters.adoc:536]:
```suggestion
=== Enable or Disable the WND In-Memory Notifier for Password Changes
```

**Review by EParzefall — APPROVED** [review]
Just some language issues.

**Comment by mmattel** [comment]
@jnweiger @jvillafanez pls have a look if the added content for #4162 is technically correct?

@EParzefall @phil-davis new content has been added at the enterprise only auth doc, can you pls have a language review on that?

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 8]
```diff
@@ -1,45 +1,188 @@
...
+    'password' => 'mypassword',
+  ],
+  'server2' => [
+    ....
+  ],
+]
+----
+
+customApp.config::
+This can be any php conform name like the above `customApp.config` or as an another example `my.config.key`, but not reserved ownCloud keys.
```
Comment (phil-davis) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:172]: I am not sure what "conform" is referring to here. Maybe you intend to say that the "name" must be a string that is valid as an array key in a PHP array? If so, than actually any string could be used (even a UTF string in a different script). And looking down further, I see that using `#` in the key string will mess things up!)

Also, I don't understand why/when a structure of nested arrays would be needed/useful. Maybe there can be a realistic example given of a system that has multiple mounts, and how that could be structured in a clear way in the config.

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 9]
```diff
@@ -1,45 +1,188 @@
...
+password::
+Each array element **must** have the key `password` present and a corresponding value, even if empty.
+Nested arrays::
+You also can use nested arrays as described in the example above. When using nested arrays, you can use as many sub-keys according your needs with any php conform name. Using nested arrays is benificial when having more than one definition but keeping it together in one master key. To access a particular subkey as config element in the mount definition, use the following scheme:
```
Comment (phil-davis) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:176]: See my comment above - maybe a "realistic" example will help make this clear.

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc (hunk 8 above)
Comment (mmattel) [reply modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:172]: > the "name" must be a string that is valid as an array key in a PHP array

This is a much better wording

> looking down further, I see that using # in the key string will mess things up!

This was the definition made by @jvillafanez (see #4162 and wnd PR) to address a particular sub-key

> when a structure of nested arrays would be needed/useful

See the issue above. The use case is, when you have more than one server (sub-key) you define but collect it in one master key. I took the description from @jvillafanez

Comment (mmattel) [reply modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc]: I have improved the text and used your comments as input. Please have a look again.
If you are fine with, pls set this block to resolved.

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 10]
```diff
@@ -1,45 +1,205 @@
...
+  Credentials are only stored in the session and not captured in the database. Files cannot be shared, as credentials are not stored.
 Log-in credentials, save in database::
   Credentials are stored in the database, and files can be shared.
 User entered, store in database::
-  Users provide their own login credentials, rather than using
-  admin-supplied credentials. User credentials are stored in the
-  database, and files can be shared.
+  Users provide their own login credentials, rather than using admin-supplied credentials. User credentials are stored in the database, and files can be shared.
+Credentials hardcoded in config file::
+  Used when all the users will share the same password and the password is not known to the users like for a shared space. The user id is always taken from the session. This authentication option can also be chosen when using oAuth or OpenIDConnect ect. Files can be shared.
```
Comment (phil-davis) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:99]:
```suggestion
  Used when all the users will share the same password and the password is not known to the users like for a shared space. The user id is always taken from the session. This authentication option can also be chosen when using oAuth or OpenIDConnect etc. Files can be shared.
```

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 11]
```diff
@@ -1,45 +1,205 @@
...
+  ],
+  'server2' => [
+    'password' => 'the_second_password',
+  ],
+  ....
+]
+----
+
+customApp.config::
+The naming of this key must be a string that is valid as an array key in a PHP array like the above `customApp.config` or as an another example `my.config.key`, but not any reserved ownCloud key.
```
Comment (phil-davis) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:172]:
```suggestion
The naming of this key must be a string that is valid as an array key in a PHP array like the above `customApp.config` or as another example `my.config.key`, but not any reserved ownCloud key.
```

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 12]
```diff
@@ -1,45 +1,205 @@
...
+
+password::
+Each array element **must** have mandatory exact one key named `password` present and a corresponding value, even if empty.
```
Comment (phil-davis) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:175]:
```suggestion
Each array element **must** have exactly one key named `password` present and a corresponding value, even if empty.
```

We don't really need to say "mandatory" when "must" is already written and highlighted.

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 13]
```diff
@@ -1,45 +1,205 @@
...
+
+Nested arrays::
+You can use nested arrays as described in the example above. When using nested arrays, you can use as many sub-keys according your needs where the naming of the sub-key must be a string that is valid as an array key in a PHP array.
++
+Note that sub-keys, exept the password key are not used as element for the mount point. Sub-keys are only present to optimize grouping and accessing the passwords.
```
Comment (phil-davis) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:180]:
```suggestion
Note that sub-keys, except the password key are not used as elements for the mount point. Sub-keys are only present to optimize grouping and accessing the passwords.
```

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 14]
```diff
@@ -1,45 +1,205 @@
...
++
+Nested arrays are benificial when having more than one server (host) with a password, but keeping them together in one master key. To access a particular sub-key as config element in the mount definition, use the following scheme:
```
Comment (phil-davis) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:182]:
```suggestion
Nested arrays are beneficial when having more than one server (host) with a password, but keeping them together in one master key. To access a particular sub-key in the mount definition, use the following scheme:
```

**Review by phil-davis — COMMENTED** [review]
The flow of text looks good now. Just a few minor editing suggestions.

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc (hunk 8 above)
Comment (phil-davis) [reply modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc]: Looks good. I just reviewed a few other minor grammar/typos.

**Comment by mmattel** [comment]
Squashed because of the language fixes.

**Review by phil-davis — COMMENTED** [review]
Generally LGTM now.

File: modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc [hunk 15]
```diff
@@ -49,16 +49,32 @@ Listen to smb changes and store notifications for later processing in the databa
...
 | `--unbuffering-option=UNBUFFERING-OPTION`
 | Force the usage of that unbuffering option for the underlying smbclient command. Possible options are either "auto", "pty" or "stdbuf" [default: "auto"]
+
 | `--output[=OUTPUT]`
 | The output format to use (plain, json or json_pretty). [default: "plain"] 
 |===
 
+[NOTE]
+====
+. To enter the password of the service account via the console, the `password` argument, the `password-file` option and the `password-from-service-account` option must not be set.
```
Comment (jnweiger) [inline modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc:71]: In this enumeration, one item is an "argument, the other two are options.
Suggest to also say ``` ... the `password` option, ...```

File: modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc [hunk 16]
```diff
@@ -49,16 +49,32 @@ Listen to smb changes and store notifications for later processing in the databa
...
+
+. To use the password for the service account from the mount point settings, use the `password-from-service-account` option.
+
+. If you use the `password` argument together with the `password-file` option, the `password-file` option takes precedence over the `password` argument.
```
Comment (jnweiger) [inline modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc:75]: Same as above: In this enumeration, one item is an "argument", the other two are "options"
```suggestion
. If you use the `password` option together with the `password-file` option, the `password-file` option takes precedence over the `password` argument.
```

File: modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc (hunk 15 above)
Comment (jnweiger) [inline modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc:71]: In this enumeration, one item is an "argument", the other two are "options".
```suggestion
. To enter the password of the service account via the console, the `password` argument, the `password-file` option and the `password-from-service-account` option must not be set.
```

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 17]
```diff
@@ -1,45 +1,205 @@
...
+| x
+| x
+|===
+
+(*) Note that `RSA public key` is only available for SFTP.
+
+== Storage Implementations
+
+The following table shows which storage implementation can use the above authentication options.
+Note that this table does not cover other storage implementations like Google Drive or One Drive ect, as those use other authentication backends.
```
Comment (jnweiger) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:51]:
```suggestion
Note that this table does not cover other storage implementations like Google Drive or One Drive etc., as those use other authentication backends.
```

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 18]
```diff
@@ -1,45 +1,205 @@
...
+  Credentials are only stored in the session and not captured in the database. Files cannot be shared, as credentials are not stored.
 Log-in credentials, save in database::
   Credentials are stored in the database, and files can be shared.
 User entered, store in database::
-  Users provide their own login credentials, rather than using
-  admin-supplied credentials. User credentials are stored in the
-  database, and files can be shared.
+  Users provide their own login credentials, rather than using admin-supplied credentials. User credentials are stored in the database, and files can be shared.
+Credentials hardcoded in config file::
+  Used when all the users will share the same password and the password is not known to the users like for a shared space. The user id is always taken from the session. This authentication option can also be chosen when using oAuth or OpenIDConnect etc. Files can be shared.
```
Comment (jnweiger) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:99]: For details see below, section "Notes for Credentials Hardcoded in Config File"

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc [hunk 19]
```diff
@@ -1,45 +1,205 @@
...
+You can use nested arrays as described in the example above. When using nested arrays, you can use as many sub-keys according your needs where the naming of the sub-key must be a string that is valid as an array key in a PHP array.
++
+Note that sub-keys, except the password key are not used as elements for the mount point. Sub-keys are only present to optimize grouping and accessing the passwords.
++
+Nested arrays are beneficial when having more than one server (host) with a password, but keeping them together in one master key. To access a particular sub-key in the mount definition, use the following scheme:
++
+[source,text]
+----
+<key>#<level_1>#<level_2>...
+----
```
Comment (jnweiger) [inline modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:187]: Hard to understand for me. What is a subkey? What is key? (probably a servername, from the exampe below) Where do levels appear?
Where exactly is this syntax used? In the entry form of the admin sharing dialog?

File: modules/admin_manual/pages/enterprise/external_storage/windows-network-drive_configuration.adoc [hunk 20]
```diff
@@ -157,7 +157,8 @@ To enable external storage, as the ownCloud administrator go to menu:Settings[St
 .When you create a new WND share, you need three things:
```
Comment (jnweiger) [inline modules/admin_manual/pages/enterprise/external_storage/windows-network-drive_configuration.adoc:157]:
```suggestion
.When you create a new WND share, you need multiple things:
```
We go from three to four here, but "credentials" was already 'two things' - in my world.

File: modules/admin_manual/pages/enterprise/external_storage/windows-network-drive_configuration.adoc [hunk 21]
```diff
@@ -170,19 +171,19 @@ if the case is not respected.
...
 . Enter the ownCloud mount point for your new WND share. This _must not_ be an existing folder.
-. Select your authentication method. See xref:enterprise/external_storage/enterprise_only_auth.adoc[Enterprise-Only Authentication Options] for complete information on the five available authentication methods.
+. Select your authentication method. See xref:enterprise/external_storage/enterprise_only_auth.adoc[Enterprise-Only Authentication Options] for complete information on the available authentication methods.
 +
 .Figure 2. WND mountpoint and authorization credentials
 image:enterprise/external_storage/windows_network_drive/wnd-2.png[WND mountpoint and authorization credentials, width=60%]
 . Enter the address of the server that contains the WND share.
-. The Windows share name.
-. The root folder of the share. This is can be the subfolder name, or the `$user` variable for the user's home directory. Note that the LDAP `Internal Username Attribute` must be set to the `samaccountname` for either the share or the root to work, and the user's home directory needs to match the `samaccountname`.
+. The share name provided by Windows .
```
Comment (jnweiger) [inline modules/admin_manual/pages/enterprise/external_storage/windows-network-drive_configuration.adoc:179]:
```suggestion
. The share name provided by Windows or Samba.
```

Suggested Note: Omit any leading `/` or `\` on the share name. A slash or backslash there may lead to odd behaviour.

**Review by jnweiger — CHANGES_REQUESTED** [review]
typos and minor notes

File: modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc (hunk 15 above)
Comment (mmattel) [reply modules/admin_manual/pages/configuration/server/occ_commands/app_commands/_wnd_commands.adoc:71]: @jnweiger the suggestion is the same as the original text. Can you adopt it?

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc (hunk 19 above)
Comment (mmattel) [reply modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc:189]: fixed, text added

File: modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc (hunk 18 above)
Comment (mmattel) [reply modules/admin_manual/pages/enterprise/external_storage/enterprise_only_auth.adoc]: fixed, link added

**Review by jnweiger — APPROVED** [review]

**Pull request merged**